src/rhsmlib/facts/cloud_facts.py

```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import print_function, division, absolute_import
+
+#
+# Copyright (c) 2019 Red Hat, Inc.
+#
+# This software is licensed to you under the GNU General Public License,
+# version 2 (GPLv2). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+# along with this software; if not, see
+# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+#
+# Red Hat trademarks are not licensed under GPLv2. No permission is
+# granted to use or replicate Red Hat trademarks that are incorporated
+# in this software or its documentation.
+#
+
+import logging
+import json
+
+from rhsmlib.facts import collector
+from rhsm.https import httplib
+
+log = logging.getLogger(__name__)
+
+
+class CloudFactsCollector(collector.FactsCollector):
+    """
+    Class used for collecting facts related to Cloud instances
+    """
+
+    def __init__(self, arch=None, prefix=None, testing=None, collected_hw_info=None):
+        super(CloudFactsCollector, self).__init__(
+            arch=arch,
+            prefix=prefix,
+            testing=testing,
+            collected_hw_info=collected_hw_info
+        )
+
+        self.hardware_methods = [
+            self.get_aws_instance_id
+        ]
+
+    def get_aws_instance_id(self):
+        values = {}
+
+        if not self._collected_hw_info['dmi.bios.version'].find('amazon') >= 0:
+            return ""
+
+        try:
+            conn = httplib.HTTPConnection('169.254.169.254', timeout=5)
+            conn.request('GET', '/latest/dynamic/instance-identity/document')
+            response = conn.getresponse()
+            output = response.read().decode('utf-8')
+            values = self.parse_content(output)
```

Comment:
There is several things that could go wrong. It will not be possible to establish connection with `169.254.169.254`. It will not be possible to get response from server. It will not be possible decode response, because encoding will not be utf-8. It will not be possible to parse content. Each case deserves own try-except and properly handling (writing warning/errors to log file). And it would be also good to have unit test for each case.